packages/react-emotion/typings/react-emotion.d.ts

```diff
@@ -0,0 +1,103 @@
+import { StatelessComponent, Component as ClassComponent, CSSProperties } from 'react'
+import { Interpolation as EmotionInterpolation } from 'emotion'
+
+export type InterpolationFn<Props = {}> = (
+  props: Props
+) => EmotionInterpolation
+
+export type Interpolation<Props = {}> =
+  | InterpolationFn<Props>
+  | EmotionInterpolation
```

Comment:
I don't know much about typescript but I don't think this handles the fact that there can be nested function interpolations and they will receive props.

```jsx
const SomeComponent = styled.div`
  display: ${(p) => props => props.display};
  color: ${[props => 'hotpink']};
`
```

Reply:
It works on emotion?
props => props_ => ...

It doesn't make any sense to me. What's this use case for?

Reply:
I don't know exactly all edge cases of emotion API. I'll try to study it to improve the coverage support.

Reply:
@mitchellhamilton I've added support to these two cases you mentioned. 

```tsx
export type InterpolationFn<Props = {}> =
  (props: Props) =>
    | EmotionInterpolation
    | InterpolationFn<Props>

export type InterpolationTypes<Props = {}> =
  | InterpolationFn<Props>
  | EmotionInterpolation

export type Interpolation<Props = {}> =
  | InterpolationTypes<Props>
  | InterpolationTypes<Props>[]
```

If you know more cases, just tell me, please.